### ISSUE TITLE
Optimising loading

### ISSUE THREAD

**Issue by JohnLukeBentley** [aaFn/Bookmark-search-plus-2#2]
@aaFn wrote in #1 
> ... about delay before showing .. Yes I started to work in the add-on on an option to immediately show the tree while it is building, asynchronously loading it. I didn't really make it good for now as it was experimental, so you see no difference if you activate it in the add-on options panel (although it really is asynchronous behind the scene).

> Now, a dialog with a guy who has 80000 bookmarks (I am a micro-small player besides him with my 1800 bookmarks lol) on the add-on review page, made me envisage a number of alternatives to really make that effective. I'll try to work out something by end of next week that should visually get more responsive .. although that won't really improve real loading perfs by themselves, just visual feeling.
Now, it is clear I won't be able to reach the load performances of native code .. no dreaming, but working on the visual interactivy + maybe on a few optimi(s/z)ations, I might be able to do something effective .. let's see.

Yes getting asynchronous loading to work sounds like the way to go. As a user you'd want to be able to start typing a search immediately. Then, the results could come in over time ... a delay before the search is complete would be far less of a problem.

Anyway it sounds like you have a plan of action in mind. All the best with that.

**Comment by aaFn** [comment]
Ok, I was toying with the idea to disable the search box wile loading .. what is sure is that displaying results while still loading can be a problem, because I don't know yet the position of the bookmark (since not yet loaded / displayed) .. so I would need at least to delay results display ..

What would you prefer ? Wait for the searchbox to be enabled, or alllow to type in, but display a waiting icon until whole load is complete and then display result (more complex, but doable I think) ?

**Comment by JohnLukeBentley** [comment]
Let's assume for the moment that displaying any results can't be done until the tree is fully loaded. In that case I'd suggest not disabling the search box. If the search box is enabled the user has something they can do, start typing.

However, if the API provides for asynchronous loading as you mention, then you should be able to load partial results. The whole point of asynchronous loading is to enable the dev to get some data without blocking current processes (or having to reload the full environment, such as a whole web page in the web page context).

So here might be an algorithm you could use:

* Fetch (and optionally store on disk for next time) a count of the total bookmarks. 
* Divide the total by a `chunkFactor`, e.g. 10, to get `bookmarkQuantiyToFetchInChunk`. If there are 1,000 total bookmarks then `bookmarkQuantiyToFetchInChunk` will be 100.
* Create a loop that asynchronously fetches the next chunk of 100 bookmarks.
* Apply the search string to the 100 bookmarks, display the bookmark(s) in the search result list, if any.
* Repeat loop until done.

Assuming that's plausible what's loaded in a chunk of bookmarks is an issue. In principle you could load (at least in the background) both a linear list for the search result pane and the tree structure for those bookmarks so far fetched. 

In practice you might be able to, or might want to for optimization reasons, only load bookmarks in the linear list. In that case you may well want to disable to the tree pane but keep enabled the search result pane. That is, you enable the tree pane once all the chunks have been loaded.

You may well already have that sort of algorithm (in the above paragraphs) in mind as a possible thing to do, were it not for the API (or other constraints).  I mention it in case it sparks a solution.

**Comment by aaFn** [comment]
Ah ah there are some good ideas in there, thanks :-)
I guess I can in fact display all results, but specially highlight the ones which are not ready yet for "Show in bookmarks" and on which clicking won't do anything in the bookmark pane + gray out the "Show in bookmarks" action in the context menu when right clicking on them.

And then maybe make it more intelligent so that as long as the result pane is open, I wake up regularly that piece to check the ones not yet enabled for show, see if they are now ready, and then enable them if yes.
I have the mechanism for that already, shouldn't be too hard to code .. :-)

I should save that for 2.0.4

**Comment by NicolasWeb** [comment]
> what is sure is that displaying results while still loading can be a problem, because I don't know yet the position of the bookmark (since not yet loaded / displayed) .. so I would need at least to delay results display ..

Is there any Firefox API or Firefox platform enhancement that could help to make the search result asynchronous loading ? That might be useful for this addon and others...
If yes, let's open a bug in Bugzilla ;-)

**Comment by aaFn** [comment]
Thinking twice on this, I am getting to the point where there is a fundamental design choice to make for the "asynchronous" feature:
- Either do as we discuss below, loading and displaying all hidden things in **background**
- Either simply not load/display what is hidden (inside closed folders), and when a folder is open (by a search or by a click on twistie), then fetch them and display them at that moment = **on-demand**

In the second case, I do not have the same complexity of intermediate state to maintain, and it on't have a "dirty background process" doing things behind the scene, so this is seducing.
The drawback is possible delay on "show bookmark" and on "open folder", when the number of things to display at those moments gets too large

-> I guess this all revolves around the ratio of avg number of bookmarks per folder .. If they are not too large, and contents is properly distributed between folders, that will go easy.

Also, a factor of decision is the time to load the full tree. I need to see stats from @vitaaliy in #3 before I can make an educated choice.

**Comment by JohnLukeBentley** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1438829 is now marked as "RESOLVED WORKSFORME"

And FF is now at 59.0.1.

Might that clear the way for you to have another go at optimization? Given all the recent excellent work you've done on some of the details about the UI that seems to clear the deck sufficiently (of course the TODO list remains long anyway) to tackle this issue of optimization. As a user the loading time strikes me as most prominent thing needing fixing.

**Comment by aaFn** [comment]
Hello @JohnLukeBentley , as a matter of fact it appears a Windows update had set my Power Management to throttling my CPU .. so the problem is cleared now since some time.

Performances for the FF query are still awful I find .. I have been thinking at ways to improve that.
My current stance is:
1. Move all bookmark loading and storing in a background page -> This should not change the initial load time on browser opening, but this should noticeably accelerate things when later on switching the sidebar between native, BSP2 and other add-ons.
However, there is a bug (as always - sigh !) on this one : FF windows which are incognito cannot access the background page, so that optimization won't work for them. They will need to remain in the old mode.
I also need to experiment on the display of the same background object in several FF windows, so when having several sidebar instances open at the same time and displaying the same object .. not sure how that behaves ..
2. Store the bookmark tree in add-on local storage, and reload from it on initial load when FF is starting. And resynchronize it in background with the FF bookmarks in case something was changed while the add-on was not active.
This should accelerate noticeably the initial load for people with high numbers of bookmarks.
However, this may slow things down for people with small numbers .. and it has the disadvanage of having the core allocated to the thread to be at 100% (cannot control that, this is the badly tuned FF query inside the API which does that ..), so I will make it an option = the one which already exists and is not effective today.
3. I will wait for the FF query improvements to revise the above ..

Let me know if you have other ideas .. this is significant work though, so I am still not ready to go into it, and it will probably be several waves.

All the more that I am getting some good suggestions in the issues to improve, and that entertains some activity which is keeping me from really focusing on the optimization for now .. :-)

**Comment by JohnLukeBentley** [comment]
Broadly I can only think of one or more of:

* Background loading with add-on local storage, as you mention;
* Asynchronous loading, as previously discussed;
* Directly accessing the native bookmark store, as previously discussed.

So it sounds like you have the right sort of things in mind to try.

**Comment by JohnLukeBentley** [comment]
Perhaps, as a temporary workaround, you could provide an option, off by default, to do the following ...

After BSP2 initially loads the tree (i.e. when BSP2 loads for the first time in a firefox session), can you simply keep this in memory, even if BSP2 closes. 

If, in the mean time, the user loads the native bookmark sidebar and adds bookmark, the two trees will get out of sync. Perhaps you could provide a "Reload" button at the top of the BSP2 to force BSP2 to reread from the native source (or else BSP2 will just reload after a Firefox restarts and BSP2 loads for the first time again).

**Comment by aaFn** [comment]
Yes, this should be my 1) above .. but it should also contains the hooks for when there is a boomark is changed / added / removed, so as to keep things in sync.
Since this is to be a background page, it will ba active as long as FF is active.

**Comment by aaFn** [comment]
Let's have hope that some improvement can happen on the front of selecting bookmarks in FF API code: https://bugzilla.mozilla.org/show_bug.cgi?id=1396364

So for now I am concentrating on step 0 which is not listed above :-)
= improve the display speed by converting all favicons to 16x16, before storing them, so that this conversion does not have to happen at display time.

Hoping I can get some significant optimization there, while waiting something on the front of "SQL select" .. let's see, I will tell here.

**Comment by aaFn** [comment]
Not sure that improvements in the API will come soon by what Marco is saying on the bug ..

Anyway, I am done with the 16x16 conversion, and this is yielding a decrease of display time of nearly 50% on my machine !!

There is a slight visual impact to some favicons, with some being a little better (smoothing when decreasing in size), some being a little degraded (not much perceptible), a few being a little different (the scaling algorithm of 2d canvas seems different than the one used by img when being displayed on screen) and the vast majority simply looking the same.

So all in all, neither positive nor negative, but very much worth the significant improvement in speed ! :-)

A migration task has been added to trigger only once 1 minute after add-on loading, and until all stored valid favicons are converted, so that this applies also to existing favicons.

This optimization 0 will be out in 2.0.16.

**Comment by aaFn** [comment]
Second pass, I found that the algortihm of Canvas 2d context drawImage() is indeed a little poor:
- the scale down algortihm is approximative, I am doing better with a simple box algorithm approach ..
- it doesn't interpret correctly the image/x-icon mime type ... it only takes the first icon of the library, which is rarely the most appropriate ! So I rewrote that part also.

Now, we have a decrease by about 50% in display duration, and no difference in all favicons I could check ..
Going with that in 2.0.16, it should be out tonight after further testing ..

**Comment by JohnLukeBentley** [comment]
Thanks for that exploration.

On 2.0.16, however, subjectively I'm not noticing much difference.

For testing this could you verify I have the right combination of options selected:

Disable favicon fetching: unticked
Include cookies to increase ... : ticked
Load bookmarks only when required ...: ticked

**Comment by aaFn** [comment]
Hello @JohnLukeBentley , the benefit is not immediate = the existing favicons first need to be migrated / converted to 16x16.
This is happening in background automatically, starting 1 minute after the add-on is made active, and happening as long as there remains something to migrate.

To verify if this is really happening, activate the traces, and look at the trace textarea contents.
If the migration already happened, it should show:
```
structureVersion: -img16
```
or else nothing after "structureVersion:".

And when the migration is still ongoing, you will see a message appearing 1 minute after add-on activation in the same trace area, saying that migration started.

As long as all favicons are not all migrated, you will see the above, and when the last favicon is migrated, you will see a message saying "migration complete". Then, next time the add-on is started, you will see the -img16 flag after "structureVersion:", and the migration activity will never trigger anymore.

Let me know if this is not what you observe .. if so, then we should open an issue and I should investigate.

**Comment by JohnLukeBentley** [comment]
Given I'm getting ` structureVersion: -img16` it looks like my migration has completed, then. Here are two traces

    PlatformOs: win
    Setting Windows variations
    structureVersion: -img16
    Load duration: 784 ms
    Display duration: 1302 ms
    fontFamily = '"Segoe UI"'
    fontSize   = '12px'
    Stats:
    ------
    Bookmarks:  4263
    Folders:    1486
    Separators: 8
    Oddities:   0
    --------------------
    Save duration: 864 ms

.

    PlatformOs: win
    Setting Windows variations
    structureVersion: -img16
    Load duration: 820 ms
    Display duration: 1260 ms
    fontFamily = '"Segoe UI"'
    fontSize   = '12px'
    Stats:
    ------
    Bookmarks:  4263
    Folders:    1486
    Separators: 8
    Oddities:   0
    --------------------
    Save duration: 888 ms

Looking at `Load Duration` there hasn't been a significant shift compared to my baseline of 2018-02-21 https://github.com/aaFn/Bookmark-search-plus-2/issues/5#issuecomment-367268458 ... over all four cases I'm between about 775ms to 850ms.

**Comment by aaFn** [comment]
Ok this is the "Display duration" that this migration is optimizing. The Load duration is linked to the bug https://bugzilla.mozilla.org/show_bug.cgi?id=1396364 and I can't do much about it in itself, apart from the workaround steps we discussed here.

Strange that we do not see any improvement on it compared to #5, this is still around 1200 - 1300 ms as before.

Unless all your bookmarks were already 16x16 ... on which I have a strong doubt.

I'll do something in 2.0.17 to monitor that, counting if we really have all at 16x16 ... and displaying in traces to verify. Could you open an issue to track that ?

**Comment by aaFn** [comment]
Hello @JohnLukeBentley , I opened a new issue to track this.
See #35 and let's pursue there if you are ok.

**Comment by aaFn** [comment]
2.0.18 is out and is making a significant move towards optimization 1).
It was large work as requiring a new internal structure representing the bookmark tree, and not just flat lists with hash indexes.

This is also accelerating a number of operation like
- copy & paste
- cut & paste
- move by drag & drop
- and create new bookmark

when there are lots of bokmarks, as a good side effect.

If all goes well, next release should see optimization step 1) out, which should accelerate switch on/off of the sidebar after initial browser start (a toolbar button is now provided as per #18, from version 2.0.18, to switch the sidebar on/off).

**Comment by JohnLukeBentley** [comment]
Thanks @aaFn. However, I'm not seeing any improvement, on 2.0.18.

    PlatformOs: win
    Setting Windows variations
    structureVersion: -img16-bnlist
    Load local store duration: 428 ms
    Tree load duration: 804 ms
    Tree build duration: 14 ms
    Display duration: 1268 ms
    fontFamily = '"Segoe UI"'
    fontSize   = '12px'
    Stats:
    ------
    Bookmarks:  4266
    Folders:    1486
    Separators: 8
    Oddities:   0
    --------------------
    Save duration: 914 ms

Did "next release" mean 2.0.19?

**Comment by aaFn** [comment]
Hello @JohnLukeBentley , so as discussed on #35, apparently you were already on 16x16, or the conversion is resulting in no gain for you in storage and display speeds ..

Strange, but:
- I see that you have the double of bookmarks compared to me
- your load from local storage and the load bookmark tree durations are 4 times my numbers
- your write to local storage duration is 10 times my number
- your constant display duration is 5 times my current duration, but a little more than 2 times my old duration before I converted to 16x16 ..
If we follow the logics, it should have been 4 times also my old number, in average.

I am on SSD, I have a 3D graphical card, and I know that FF tries to use it when it can (hardware graphic acceleration https://support.mozilla.org/en-US/kb/upgrade-graphics-drivers-use-hardware-acceleration) .. but turning 3D off in FF on my side does not change anything to my bookmarks tree display numbers.

=> I would attempt to say:
- your processor is 2 times slower than mine (= x4 duration / x2 number)
- you are not on SSD (write is x5 = x10 duration /x2 number)
- and FF is using your graphic card while not mine, which would explain the constant duration and the fact that its duration was not 4x times mine before conversion.

Anyway, this is not going to change anything to the result, so ..

Yes, 'next release' should mean 2.0.19 .. if all goes well .. we'll see.
Thank you, aaFn.

**Comment by aaFn** [comment]
Thanks for that.
My CPU is 12063 on the same scale, so that is not x 2 ... This is merely x 1.33 at best. Maybe memory speed does the rest .. I am DDR4 2400 MHz.
My SSD is also a 850, but an EVO one, not PRO. Not sure that makes much of a difference though.
My graphic card is also a Nvidia GTX. but as I said, when I disable hardware graphix acceleration in FF options, that makes no difference, so it is not used for displaying the favicon images.

The tree traversal is recursive, so it is linear, and building the display is a simple flat table, so also linear. Maybe the FF algorithm to draw table rows is not so linear .. ?

The fact that you had no improvement on both storage and display makes me deduce only 2 possibilities:
1. You were already everywhere on SVG favicons or 16x16 favicons .. highly improbable, but cannot be excluded
2. You do not have any favicon because for some reason the fetch never worked (or because you have the "No favicon fetching" option enabled but you confirmed that not on #35), and so you get for all the same "nofavicon.png" image, which is already 16x16 since the beginning.

Out of curiosity, if you do a right click -> Refresh favicon on any of your bookmarks, what is happening ? Any error in the browser console ? If no problem and favicons are all here, then we can only deduce 1) ...

Cheers.

**Comment by aaFn** [comment]
So that won't be in 2.0.19 .. I just found a blocking bug in the new structure when a bookmark is created while BSP2 is not active ... need to distribute the correctoin asap .. emergency fix :-)

**Comment by andreygursky** [comment]
I've just found this extension as replacement of the extension "Sidebar Bookmarks Search Plus". Great!

But it doesn't use the Firefox's native UI for bookmarks as the mention extension, which worked instantly. Instead is does (must?) use custom UI. This results in huge performance drawback (loading times, memory consumption, less responsive browser).

My data:
```
PlatformOs: linux
Setting Linux variations
structureVersion: -img16-bnlist
disablefavicons_option: false
Load local store duration: 5546 ms
BSP2 version: 2.0.19
Tree load duration: 29544 ms
Tree build duration: 944 ms
Display duration: 14186 ms
fontFamily = 'Sans'
fontSize   = '12px'
Stats:
------
Bookmarks:  51156
Folders:    64
Separators: 1
Oddities:   0
--------------------
Save duration: 4964 ms
```

The main question: is it not possible anymore to create only custom widget for search field and results field (thus avoiding loading all bookmarks into UI completely) but manipulate the extremely efficient Firefox's bookmarks widget (move to the selected found bookmark)?

Thanks for keeping working on it!

**Comment by aaFn** [comment]
Hello @andreygursky , unfortunately, the only left mode for add-ons in FF as of version 57+ is what is called WebExtensions, and we cannot interact anymore with FF native objects / widgets, for security reasons.
We must go through API's https://developer.mozilla.org/en-US/Add-ons/WebExtensions
and the ones to manipulate bookmarks are less than optimal ..

With the number of bookmarks you have, the simple getTree() query https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/bookmarks/getTree takes nearly 30s from your stats ..
This is just to get the bookmarks .. and this is internal to the API, nothing we can do about .. https://bugzilla.mozilla.org/show_bug.cgi?id=1396364

The search query goes faster, but since there is no way to interact with the native UI tree, there is no way to show the corresponding bookmark so far ...

Sorry for the bad news ... and I am eager like you to find better ways .. or just to wait for them to come https://bugzilla.mozilla.org/show_bug.cgi?id=1420246

**Comment by andreygursky** [comment]
@aaFn, thanks for links!

https://bugzilla.mozilla.org/show_bug.cgi?id=1396364#c5
>The expensive part of the SQL query in promiseBookmarksTree appears to be the tags extraction for each item...

https://addons.mozilla.org/en-US/firefox/addon/bookmark-search-plus-2/
> There are some limitations though compared to the old extension, because of Web extension security restrictions: cannot ... display of bookmark tags

So let's spend a lot of time extracting something that will then dropped and not made available for users. Strange. If Web Extensions are still not really ready for production use, why the NPAPI has been dropped already...


https://bugzilla.mozilla.org/show_bug.cgi?id=1396364#c7
> tags extraction slows down a lot of queries, we can't do anything for now. The good thing is that a minority of users has many bookmarks, and a minority of that minority has tags...

Oh, then I must be lucky I haven't yet started to use such useful thing as tags.


https://bugzilla.mozilla.org/show_bug.cgi?id=1420246#c6
> The problem is that we don't yet have an efficient way to get the full path to a bookmark 

But the old extension worked instantly... So what is the (new) problem?


Showstopper "tags should not be bookmarks folders, move them to a separate table" https://bugzilla.mozilla.org/show_bug.cgi?id=424160
>  Reported: 10 years ago (20.03.2008)

That's really sad.

**Comment by aaFn** [comment]
I believe the query used by the native code is not the same as the one given to extensions .. Or they have a form of caching + getting deltas which avoid them to use the query .. I will see if I can approach that by using a background task always present, which would avoid me reusing getTree() at each start, and which would in your case reduce the total duration to:
```
Load local store duration: 5546 ms
Tree build duration: 944 ms
Display duration: 14186 ms
```
so a total of around 20s. Not ideal, but always better than a total of 50s ..
Only constraint would be for the user to always leave the add-on active in background (i.e. not uninstall / deactivate it) or else he/she will incur the API penalty at next activation / install ..
Probably acceptable.

Next big thing would be display, but in WebExtensions I am reduced to manipulate only DOM trees and HTML, while in the native sidebar I believe they use native OS objects, much faster ..
So as discussed in this thread with others, I would have to switch to a width first algorithm when reaching a closed folder, leaving the closed folder for later background display .. That would make the display phase seemingly faster by concentrating only on what is visible first (of course if the user expands the whole tree, no benefit ..)

Slow work as all these workarounds are somewhat complex .. :-)

**Comment by JohnLukeBentley** [comment]
@aaFn, just to report my stats on v2.0.19

    PlatformOs: win
    Setting Windows variations
    structureVersion: -bnlist-img16
    disablefavicons_option: false
    Load local store duration: 474 ms
    BSP2 version: 2.0.19
    Tree load duration: 878 ms
    Tree build duration: 16 ms
    Display duration: 1419 ms
    fontFamily = '"Segoe UI"'
    fontSize   = '12px'
    Stats:
    ------
    Bookmarks:  4270
    Folders:    1487
    Separators: 8
    Oddities:   0
    --------------------
    Save duration: 1039 ms

Compared to the `Display duration: 1268 ms` in my last stats ... it's gotten worse. 

> I would also need a way to detect that I am out of sync

Can you retrieve a count of bookmarks in the native store? Comparison to the count of bookmarks in the BSP2 could give you a crude out-of-sync signal.

Even if that is not possible, how about the following workaround, until out-of-sync detection is available? ...

Give us an option "For speed allow BSP2 store and native store to drift out of sync in a session". If ticked:
* On first load of FF (for every session) and a subsequent open of BSP2, BSP2 loads as normal.
* Closing the panel BSP2 only closes the UI, the BPS2 store is kept in the background.
* In the same session a subsequent opening of BSP2 should, therefore, be near instantaneous.
* On close of FF BSP2 writes to the native store (instead of every time the BSP2 panel is closed).

That way sync occurs at the start and end of every FF session, but during the session BSP2 is fast.

If FF crashes, then you'll lose bookmark changes for that session. But that'd be part of the risk in ticking the option.

(Or maybe that's your plan anyway).

Good luck in the struggle.

**Comment by aaFn** [comment]
Hello @JohnLukeBentley , yes I was thinking of something like what you say =
- next release N, use native store on FF start to always be sync'ed, then only use local store for the whole FF process life. Keep writing to both native and local stores on each bookmark update as 1) this does not bear perf problems, 2) this guarantees no problem in case of crash and 3) (most of all) this allows a neat mechanism of replicating in every BSP2 session the same modifications, real time, without complex messaging of my own, by relying on native events.
- release N+1, when we are sure everything with local storage works well, provide an option to start from native every time, but untick it, and by default start from local storage. That option will be useful in case something goes wrong only, to force re-sync.

The bookmark count is not available except by the long query .. so that is not an option. And anyway, it is not reflecting updates (to names or URL's, not changing bookmark count).

_Last on display perfs:_
- I guess like me you upgraded to FF 60.0 ?
- I saw a neat decrease in display perfs on that FF release on my system also, **by approx 20% !!**
- This is not related to 2.0.19, because I developed it on 59.x, and perfs were like before.
- The only thing which changed when I saw the perf decrease was FF version.
- Now, I just upgraded tonight to 60.0.1, and guess what .. display perfs are back to the numbers before 60.0 :-)

So I guess there was something bad in 60.0, and happily they saw it and corrected it ... :-)

**Comment by JohnLukeBentley** [comment]
Going backwards through the issues. 

On current performance ...

This is morning I was on FF 60.0. But I upgraded to FF 60.0.1. On 60.0.1 the stats are

    PlatformOs: win
    Setting Windows variations
    structureVersion: -bnlist-img16
    disablefavicons_option: false
    Load local store duration: 5383 ms
    BSP2 version: 2.0.19
    Tree load duration: 1039 ms
    Tree build duration: 16 ms
    Display duration: 1396 ms
    fontFamily = '"Segoe UI"'
    fontSize   = '12px'
    Stats:
    ------
    Bookmarks:  4271
    Folders:    1487
    Separators: 8
    Oddities:   0
    --------------------
    Save duration: 1409 ms

... so no significant change, looking at Display duration. But this is a mere curiosity for there's plan ....

On your optimization plan ... that sounds much better than my idea. The clever thing in your proposal is that it avoids using an out-of-sync signal (at least as a frequent check).

> provide an option to start from native every time (keeping release N behavior), but untick it, and by default start from local storage. That option will be useful in case something goes wrong only, to force re-sync.

That's a good idea. But might it be better presented as a Button (in Options) rather than a checkbox? That is, when you click the button the sync occurs immediately (rather than forcing sync on next restart). 

In addition might we want another button (in Options) that performs an out-of-sync check, even though that's going to be an expensive operation. Because it would be an (advanced) user initiated operation .... they'll be prepared to purchase the slow down for the gain of knowing everything is OK.

**Comment by aaFn** [comment]
> But might it be better presented as a Button (in Options) rather than a checkbox?

Indeed, good suggestion, a button like the `Re-trigger 16x16 favicon migration` one, for a one time re-sync in addition to the checkmark option, enabled only when the option is not set, would make good sense, thank you.
Given the way the background algorithm works in the code, it would work at next restart like all options .. Indeed, if I want to do it real time, I have to write a diff algorithm for trees to identify the nature of modifications (move, change, delete or insert) and that is not a simple one.

Same for the check, this would require the same tree diff algorithm .. An interesting mathematical and algorithmics problem, but I would need some time to explore the litteracy on the subject and to solve it. So I'll keep it for quiet times :-)

On perfs, I notice above your Tree load duration is longer than usual, and same for the Load local store duration .. So there might have been something running in the background, or some cache not ready on that run. I would suggest you do a few recycling (by pressing on the button in the toolbar if you kept it, or several Ctrl-Q), to get more stable stats.

**Comment by aaFn** [comment]
Update: now all optimization steps discussed here are implemented as part of 2.0.27 and 2.0.28:

1. Use native store on FF start to always be sync'ed, then only use local store for the whole FF process life,
2. Use local storage on start also, and provide an option to start from native every time, but untick it, and by default start from local storage.

They are yielding significant improvements as expected, so I am closing this issue, and will wait for the FF query improvements now (https://bugzilla.mozilla.org/show_bug.cgi?id=1396364), before coming back on all this.

**Note**: the width first algorithm we briefly discussed above is a big effort as it will require background display with task execution reordering depending on where the user clicks.
It would be an interesting algothmics challenge, but it will deserve its own issue given its complexity if we ever want to go there.

So this is the end of this issue. Thank you, aaFn.

**Comment by JohnLukeBentley** [comment]
Hi aaFn,

    PlatformOs: win
    Setting Windows variations
    Load local store duration: 13 ms
    Background load local store duration: 282 ms
    Background bypassed FF API for tree load
    Background tree load duration: 4 ms
    Background tree build duration: 3 ms
    Background save duration: 126 ms
    Get tree duration: 0 ms
    structureVersion: -bnlist-img16
    disableFavicons_option: false
    Display duration: 285 ms
    Total duration: 298 ms
    fontFamily = '"Segoe UI"'
    fontSize   = '12px'
    Stats:
    ------
    Bookmarks:  4290
    Folders:    1497
    Separators: 8
    Oddities:   0
    --------------------
    BSP2 version: 2.0.28
    Favicon display duration: 1740 ms

So about a 4 to 5 x speed improvement. Subjectively much faster. Outstanding work there! Thanks very much.

It would be good to get another 2 x speed improvement to get total duration under 150ms ... hopefully the FF bug, whenever it drops, will give us that.

But, again, great work on the current optimization push!

**Comment by aaFn** [comment]
Happy this is good for you also then :-)
Thank you for you support, aaFn.

**Comment by andreygursky** [comment]
I've just restarted Firefox 61.0. After I clicked on History -> Restore Previous Session, Firefox shows strange behavior. It is normal that for a couple of seconds it becomes unresponsive until all windows/tabs get restored and memory consumption grows, but then it drops again to a reasonable level and Firefox is ready to use.

Now I'm observing that **idle** Firefox keeps downloading something from many different sites. Memory consumption keeps growing (sometimes 100% of RAM, so that swapping starts) and dropping.

![firefox-memory consumption](https://user-images.githubusercontent.com/280321/42567702-e180c28e-850a-11e8-8b77-e2a149284891.png)

Firefox is unusable. @aaFn, have you enabled some background task by default?

**Comment by andreygursky** [comment]
It is not necessary to restore the session, just wait a bit after start to trigger the described behavior.

**Comment by andreygursky** [comment]
Once I've disabled favicon fetching and restarted Firefox, all seems to be OK. I've opened a new issue: https://github.com/aaFn/Bookmark-search-plus-2/issues/49.

**Comment by JohnLukeBentley** [comment]
Just to present the latest stats since the update to 2.0.29 :

    PlatformOs: win
    Setting Windows variations
    Load local store duration: 12 ms
    Background load local store duration: 741 ms
    Background bypassed FF API for tree load
    Background tree load duration: 98 ms
    Background tree build duration: 14 ms
    Background save duration: 852 ms
    Get tree duration: 1 ms
    structureVersion: -bnlist-img16
    disableFavicons_option: false
    pauseFavicons_option: false
    Display duration: 289 ms
    Total duration: 302 ms
    fontFamily = '"Segoe UI"'
    fontSize   = '12px'
    Stats:
    ------
    Bookmarks:  4297
    Folders:    1499
    Separators: 8
    Oddities:   0
    --------------------
    BSP2 version: 2.0.29
    Favicon display duration: 1587 ms

So total duration is essentially the same as last time (302ms V  298 ms). I was not necessarily expecting an improvement. However, subjectively it felt faster since the upgrade to 2.0.29. Evidently that was a trick of the mind ... but since the subjective experience is what matters I'll happily take the placebo :).

**Comment by aaFn** [comment]
Placebo are good also :-D
Clearly, the step s we devised are much better than before anyway, so they were worth the effort.
I will get back to cosmetic things now ... Thanks, aaFn.
